### PR TITLE
Handle Monorepo Releases

### PR DESCRIPTION
**Pull request by janl**
This PR is a follow up for #711.

This is a first sketch.

### Overview

The overall goal of this feature is an improvement to the Greenkeeper User Experience when depending on packages that are part of a monorepo. At the moment, users receive a separate update notification for each sub-package in a monorepo. This is less than ideal, because generally, without updating all packages to the same release version, the dependency update doesn’t work fully, so users have to go through `sub-packages.length` potentially broken CI builds/PRs before they have updated from a monorepo dependency.

We want to change this so updates to packages that are part of a monorepo are handled in unison. For example, package `wibble-wobble` consists of two sub-packages `wibble-wobble-core` and `wibble-wobble-extra` that are always released together and on the same version number. With this update, we’d like Greenkeeper to open a *single* issue/branch/PR that includes the `package.json` update for *both* packages.

### How it works

Greenkeeper updates go through two jobs:

- `registry-change` handles update notifications from npm (both the public registry as well as from web  hooks for private packages). This runs once per version update.
- `create-version-branch` creates a branch for a monitored repository that has the dependency that is handled in `registry-change`. This runs for *every* repo that is enabled for our users.

We are changing these with the help of code in `lib/monorepo.js` to handle the above scenario. In addition `util/monorepo-definitions.js` includes a static list of the structure of monorepos. With that, we can decide if a given `registry-change` is:

  1. part of a monorepo, and
  2. the *last* of the updates required for a combined update.

If it is not the last, `registry-change` simply exits and waits for a future incarnation to issue the actual `create-version-branch`-es. In addition, it writes a monorepo release info doc into the `npm` database that includes the timestamp of the latest monorepo release that didn't cause any `create-version-branch` jobs to be created.

This, together with a view that sorts these docs by timestamp, we are introducing one more component: the `monorepo-supervisor`. It runs periodically (every 5 minutes) and checks if the aforementioned view returns any documents for the last >= 30 minutes. If it finds any, it re-issues the last `registry-change` job again, with a `force: true` parameter, that makes `registry-change` not bail out. In addition, we receive a Slack/Email notification, that a new monorepo release happened and we can check if our static definition is still up to date.

The following failure scenarios for out of date monorepo definitions are considered:
- a new package was added:
  - this is not a failure, because being a new package, nobody has a dependency on it yet, so we don’t have to send updates.
  - however, we should update our monorepo definition because the *next* update will require us to do updates.
- a package was deleted:
  - without the `monorepo-supervisor` job, we’d wait for this update forever.
  - we also get a notification so we can update our definition
- a package was renamed:
  - from our point of view, this is a package delete and an unrelated add, so we can’t really offer a good update experience. We can still create a branch so users can do the fix manually (this is no different from today).

`create-version-branch` will require a few modifications, too: instead of handling a single package update, it now needs to detect a monorepo and update *all* dependencies that are updated by the monorepo release for the current package. It will create a single `package.json` transform with all required changes. The same will have to be done to `create-group-version-branch`.

* * *

We want to make updating the monorepo definitions a matter of updating our database, so that we can do the update without full stage/production deploy cycles.

### Enterprise?

There are three possible scenarios:

1. Enterprise installation with [DMZ](https://en.wikipedia.org/wiki/DMZ_(computing)) access and lazy admin

Here, we can add a periodic auto-updater that fetches the latest monorepo definitions from our service.

2. Enterprise installation with DMZ access and eager admin

Here, we can notify the admin via email, that a monorepo definition might require a manual update, and we’ll explain how to make the edit in the GKE database. Even allow definition of private monorepo structures.

3. Enterprise installation **without** DMZ access

Like 2, and no updates.

* * *

## ToDo:
- [x] Write More Tests
- [x] Handle updates in create[-group]-version-branch
- [x] adapt lib/get-infos
- [x] adapt lib/get-diff-commits
- [x] adapt lib/handle-branch-status
- [x] Add the monorepo definitions to the database for easier updating, and customisation for Enterprise users.
- [x] Add Slack notifications
- [x] ~Add Email notification~ not in this PR
- [x] ~decide whether and how we want to handle monorepos that do not assign the same version number to all packages.~ -> not in this PR
- [x] figure out all places after `create[-group]-version-branch` that need adjustment, like branch-cleanup etc.
- [x] decide whether the release update info doc gets deleted in `create-version-branch` or in `registry-update`
- [x] discuss: reduce wait period to 5 or 10 minutes

### CODE REVIEW

File: jobs/registry-change.js [hunk 1]
```diff
@@ -65,6 +72,22 @@ module.exports = async function (
     return
   }
 
+  let dependencies = [ dependency ]
+
+  // check if dependency update is part of a monorepo release
+  if (isPartOfMonorepo(dependency)) {
+    const version = distTags['latest']
+    if (force === true || !await hasAllMonorepoUdates(dependency, version)) {
+      log.info('exited: is not last in list of monorepo packages')
```
Comment (espy) [inline jobs/registry-change.js]: Shouldn’t it be `force === false`? We _want_ this to continue if it is forced, right?

File: lib/monorepo.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    if (e.message !== 'not_found') {
+      throw e
+    }
+  }
+}
+
+async function pendingMonorepoReleases () {
+  const { npm } = await dbs()
+  // thx https://stackoverflow.com/questions/1197928/how-to-add-30-minutes-to-a-javascript-date-object#1214753
+  const thirtyMinutesAgo = () => new Date(new Date().getTime() + 30 * 60000).toJSON()
```
Comment (espy) [inline lib/monorepo.js]: This is 30 minutes in the future, should be `new Date(new Date().getTime() - 30 * 60000).toJSON()`, no?

File: jobs/create-group-version-branch.js [hunk 3]
```diff
@@ -44,6 +55,26 @@ module.exports = async function (
...
   const log = Log({logsDb: logs, accountId, repoSlug: repository.fullName, context: 'create-group-version-branch'})
   log.info('started', {dependency, version, oldVersion})
+
+  // if this dependency is part of a monorepo suite that usually gets released
+  // all at the same time, check if we have update info for all the other
+  // modules as well. If not, stop this update, the job started by the last
+  // monorepo module will then update the whole lot.
+  if (await isPartOfMonorepo(dependency)) {
+    isMonorepo = true
+    if (!await hasAllMonorepoUdates(dependency, version)) {
```
Comment (Realtin) [inline jobs/create-group-version-branch.js]: this has to be removed as well

Comment (janl) [reply jobs/create-group-version-branch.js]: on it

**Comment by travi** [comment]
great work on this. this has really cleaned up the process that used to require a fair amount of manual handling because of conflicts with each package being so close together in the package file(s).

keep up the great work!

**Comment by janl** [comment]
Thank you @travi <3 :)

**Pull request merged**